README.md
# IPK Project 1: IPK Calculator Protocol
**Title**: IPK Calculator Protocol\
**Author**: Elena Marochkina\
**Language**: en-US\
**Supported OS**: Linux\
**Gitea**: [xmaroc00](https://gitlab.fit.cvut.cz/emarochk/ipk-proj1)

# Table of Contents
1. [Description](#description)
   1. [Compilation](#compilation)
   2. [Usage](#usage)
   3. [Included libraries](#included-libraries)
   4. [Defined constants](#defined-constants)
   5. [Global variables](#global-variables)
   6. [Defined functions](#defined-functions)
   7. [Main function](#main-function)
2. Testing
   1. [Testing the TCP client](#testing-the-tcp-client)
   2. [Testing the UDP client](#testing-the-udp-client)
3. [License](#license)
4. [Bibliography](#bibliography)

## Description

This repository contains an app which implement a client for the IPK Calculator
Protocol. The client is able to communicate with any server using IPKCP.

IPKCP is an application protocol for communicating with an Arithmetic as a Service (AaaS) provider.

## Compilation
### Linux
Before using the client, you need to compile the client.c file.

To compile the client.c file, you can run **make build**, and to remove the ipkcpc executable 
file, you can run **make clean**.

### Windows
The program can be compiled for Windows.

## Usage
The client application requires three arguments to run:

- **-h host**: specifies the hostname of the server to connect to.
- **-p port**: specifies the port number of the server to connect to.
- **-m mode**: specifies the protocol to use for communication, either tcp or udp.


      Usage: ./ipkcpc -h host -p port -m mode

## Included libraries
The code uses the following libraries:

- stdio.h - for standard input/output operations.
- stdlib.h - for general purpose functions such as memory allocation and string conversions.
- unistd.h - for POSIX API functions like close().
- getopt.h - for parsing command-line arguments.
- string.h - for string manipulation functions such as strcmp().
- sys/socket.h - for socket related functions.
- netinet/in.h - for Internet address family (IPv4 and IPv6) related functions.
- arpa/inet.h - for IP address conversion functions.
- signal.h - for signal handling functions.

## Defined Constants
**BUF_SIZE** - an integer value to specify the maximum size of the message buffer.

## Global Variables
**buffer** - a character array used to store the message buffer.\
**SOCKET** - an integer value used to store the socket file descriptor\
**MODE** - a character pointer used to store the communication protocol mode

## Defined functions

1. **void validate(char\* host, int port)** - function to validate the command-line
   arguments.

*Input Parameters*
- host - a character pointer to the server hostname.
- port - an integer to the server port number.

*Output*
- The function doesn't return anything, but it terminates the program with an error message 
if any of the arguments is invalid.

2. **void create_socket()** - function to create a socket and connect to the server.

*Input Parameters*
- MODE - a character pointer to the communication protocol mode.
- SOCKET - an integer to the socket file descriptor.
  
*Output*
- The function doesn't return anything, but it creates a TCP socket using the socket function 
with the AF_INET domain, SOCK_STREAM type, and protocol value of 0 or
  a UDP socket using the socket function with the AF_INET domain, SOCK_DGRAM type, and protocol
value of 0.
- If the socket creation fails for either case, the function prints an error message using 
the perror function and exits the program with a failure status code using the exit function

3. **void timeout()** - function to set a timeout value for the select() function when 
using UDP or TCP.

*Input Parameters*
- SOCKET - an integer to the socket file descriptor.

*Output*
- The function doesn't return anything, but it sets the timeout to 60 seconds and waits for 
- incoming data on the socket using the select() function.
  If no data is received within the timeout period, it prints an error message and exits
  the program.

4.**void tcp_client(char\* host, int port)** - function to communicate with the server using TCP.

*Input Parameters*
- host - a character pointer to the server hostname.
- port - an integer to the server port number.

*Output*
- The function doesn't return anything, but it connects to the server using the connect() 
function and sends a message to the server using the send() function.\
  It then waits for a response from the server using the recv() function and prints the
  response to the console.
This loop continues until the user types "BYE\n". Finally, it closes the socket using the 
close() function.

5. **void udp_client(const char\* server_ip, int server_port)** - function to communicate with the server using UDP.

*Input Parameters*
- host - a character pointer to the server hostname.
- port - an integer to the server port number.

*Output*
- The function doesn't return anything, but it sends a message to the server using the sendto()
function and waits for a response from the server using the recvfrom() function.\
  This loop continues until the user terminates the program with Ctrl-C.\
  Finally, it closes the socket using the close() function.

6. **void sigint_handler(int signum)** - function to handle the SIGINT signal.

*Input Parameters*
- signum - an integer to the signal number.

*Output*
- The function doesn't return anything, but it checks if the buffer is not empty, 
it clears the buffer using the memset function.\
If the MODE is "tcp", it sends the string "BYE\n" and waits for a response from the
server using the recvfrom function. \
Finally, the function closes the socket using the close function and exits the program 
with an exit code of 0.

## Main function
The main function takes two arguments **argc and argv[]**.

The main function *registers a signal handler* for SIGINT signal, which is generated when 
the user presses **Ctrl-C** on the terminal to stop the execution of the program.

It also **parses the command line arguments** using the **getopt** function. The arguments 
that it looks for are **-h** for host, **-p** for port, and **-m** for mode.\
These arguments are 
optional, but if they are specified, their values are stored in the **host, port, and MODE** 
variables respectively.

If any invalid argument is provided or any required argument is missing, the program 
will print an error message and exit.

Then, the program calls the **validate function** with **host** and **port** as arguments to validate 
if they are valid. After that, the program creates a socket using the **create_socket function**.

Finally, depending on the MODE, the program calls either the **tcp_client** or **udp_client** 
function with host and port as arguments to create a client connection.

The main function returns 0 to indicate successful execution of the program.

## Testing
### Testing validation
The program was tested on the FIT VUT Linux server using the following commands:

![IMAGE_DESCRIPTION](./Testing_pics/Validation_1.png)

### Testing for TCP client
The program was tested on the FIT VUT Linux server using the following commands:

![IMAGE_DESCRIPTION](./Testing_pics/TSP_1.png)
![IMAGE_DESCRIPTION](./Testing_pics/TSP_2.png)

### Testing for UDP client
The program was tested on the FIT VUT Linux server using the following commands:

![IMAGE_DESCRIPTION](./Testing_pics/UDP_1.png)

# License
This project is licensed under the GNU General Public License - see the 
[LICENSE](https://git.fit.vutbr.cz/xmaroc00/IPK_Project1_2023/src/branch/master/LICENSE.md)
file for details.

## Bibliography

1. [Open Group Library](https://pubs.opengroup.org/)
2. [NESFIT - IPK-Projekty - Stubs](https://git.fit.vutbr.cz/NESFIT/IPK-Projekty/src/branch/master/Stubs)
3. [Stack Overflow](https://stackoverflow.com/questions/4217037/catch-sigint-signals-in-c)
4. [Wikipedia: Markdown](http://wikipedia.org/wiki/Markdown)

CHANGELOG.md
# IPK Project 1: IPK Calculator Protocol
**Title**: IPK Calculator Protocol\
**Author**: Elena Marochkina\
**Language**: en-US\
**Supported OS**: Linux\
**Gitea**: [xmaroc00](https://gitlab.fit.cvut.cz/emarochk/ipk-proj1)

# Brief Description
The program allows the user to send and receive messages to a server using either the TCP 
or UDP protocol. The program takes in the host name, port number, and mode (TCP or UDP) 
as command-line arguments. It validates the arguments to make sure they are valid, creates 
a socket based on the protocol mode, and then sends and receives messages to and from the server.

The program includes several functions that perform different tasks. 
The validate() function validates the command-line arguments to ensure that all required 
arguments are entered, the port is valid, the mode is either TCP or UDP, and the host is 
a valid IP address.

The create_socket() function creates a socket for the specified mode (TCP or UDP) and sets 
the global variable SOCKET to the socket file descriptor.

The timeout() function sets a timeout for the select() function to prevent the program from 
waiting indefinitely for a response.

The tcp_client() and udp_client() functions connect to the server and sends a message using TCP or UDP. 
The functions reads input from the user and sends it to the server using send(). 
It then waits for a response from the server using recv(). 

The program includes a main() function that parses the command-line arguments using the 
getopt() function, calls the validate() and create_socket() functions, and then calls either 
the tcp_client() or udp_client() function depending on the specified mode.

# Known Limitations
- Limited error handling: While the program does check for some invalid inputs and errors, 
such as invalid port numbers and host names, it does not handle all possible errors that 
could occur during runtime. For example, if the server is not running or if the connection 
is lost during the communication process, the program will terminate abruptly without 
providing a clear error message.
- Limited functionality: The program is designed to send and receive messages from a 
server using either TCP or UDP protocols. It does not provide any other functionality, 
such as file transfers or remote execution of commands.
- Limited platform compatibility: The program is written in C and relies on standard 
system libraries for socket communication. This means that it may not be compatible 
with all platforms or operating systems, especially those with non-standard networking 
stacks.
- Messages larger than 1024 bytes will be truncated

client.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <signal.h>


#define BUF_SIZE 1024

//Global variables
char buffer[BUF_SIZE];
int SOCKET = 0;
char *MODE;

//Function to validate arguments
void validate(char* host, int port){

    //check if all arguments are entered
    if (host == NULL || port == 0 || MODE == NULL) {
        fprintf(stderr, "Usage: %s -h host -p port -m mode\n");
        exit(EXIT_FAILURE);
    }

    //check if port is valid
    if (port < 1024 || port > 65535) {
        fprintf(stderr, "Invalid port number. Please enter a port number between 1024 and 65535\n");
        exit(EXIT_FAILURE);
    }

    //check if mode is valid
    if (strcmp(MODE, "tcp") != 0 && strcmp(MODE, "udp") != 0) {
        fprintf(stderr, "Invalid mode. Please enter tcp or udp\n");
        exit(EXIT_FAILURE);
    }

    //check if host is valid
    struct sockaddr_in sa;
    int result = inet_pton(AF_INET, host, &(sa.sin_addr));
    if (result == 0) {
        fprintf(stderr, "Invalid host. Please enter a valid host\n");
        exit(EXIT_FAILURE);
    }
}

//Function to create socket
void create_socket () {
    if (strcmp(MODE, "tcp") == 0) {
        SOCKET = socket(AF_INET, SOCK_STREAM, 0);
        if (SOCKET < 0) {
            perror("socket");
            exit(EXIT_FAILURE);
        }
    }
    else if (strcmp(MODE, "udp") == 0) {
        SOCKET = socket(AF_INET, SOCK_DGRAM, 0);
        if (SOCKET < 0) {
            perror("socket");
            exit(EXIT_FAILURE);
        }
    }
}


//Function for timeout when using UDP
void timeout(){
    // set the timeout value for the select function
    struct timeval timeout;
    timeout.tv_sec = 60;   // 60 seconds
    timeout.tv_usec = 0;

    // monitor the socket for incoming data or timeout
    fd_set read_fds;
    FD_ZERO(&read_fds);
    FD_SET(SOCKET, &read_fds);
    int select_result = select(SOCKET+1, &read_fds, NULL, NULL, &timeout);
    if (select_result == 0) {
        printf("ERR: Timeout\n");
        close(SOCKET);
        exit(1);
    }
}

//Function to send message to server and receive answer using TCP
void tcp_client(const char* server_ip, int server_port) {
    struct sockaddr_in server_addr;

    // Set server address
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(server_port);
    if (inet_pton(AF_INET, server_ip, &server_addr.sin_addr) <= 0) {
        perror("inet_pton");
        exit(EXIT_FAILURE);
    }

    // Connect to the server
    if (connect(SOCKET, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        perror("connect");
        exit(EXIT_FAILURE);
    }

    // Send a message to the server
    while(strcmp(buffer, "BYE\n") != 0) {
        memset(buffer, 0,BUF_SIZE);
        fgets(buffer, BUF_SIZE, stdin);
        if (send(SOCKET, buffer, strlen(buffer), 0) < 0) {
            perror("send");
            exit(EXIT_FAILURE);
        }

        // Receive a response from the server
        memset(buffer, 0, sizeof(buffer));
        timeout();
        if (recv(SOCKET, buffer, sizeof(buffer), 0) < 0) {
            perror("recv");
            exit(EXIT_FAILURE);
        }

        printf("%s\n", buffer);
    }
    // Close the socket
    close(SOCKET);

}

//Function to send message to server and receive answer using UDP
void udp_client(const char* server_ip, int server_port) {

    // Set up the server address
    struct sockaddr_in server_addr;
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_addr.s_addr = inet_addr(server_ip);
    server_addr.sin_port = htons(server_port);

    // Prompt the user for a message
    while (1) {
        memset(buffer, 0, BUF_SIZE);
        fgets(buffer, BUF_SIZE, stdin);

        //add \0 to the start of buffer
        char new_buffer[BUF_SIZE];
        new_buffer[0] = 0;
        new_buffer[1] = strlen(buffer);
        memcpy(new_buffer + 2, buffer, strlen(buffer));

        // Send the message to the server
        int len = sendto(SOCKET, new_buffer, strlen(buffer) + 2, 0, (struct sockaddr *) &server_addr, sizeof(server_addr));
        if (len < 0) {
            perror("ERR: Failed to send message");
            close(SOCKET);
            exit(1);
        }

        // Receive the server's response
        memset(buffer, 0, sizeof(buffer));
        timeout();
        len = recvfrom(SOCKET, buffer, BUF_SIZE, 0, NULL, NULL);
        if (len < 0) {
            perror("ERR: Failed to receive response");
            close(SOCKET);
            exit(1);
        }

        int status_code = (int) buffer[1];
        if(status_code == 0){
            printf("OK: %s\n", buffer + 2);
        } else {
            printf("ERR: Invalid expression\n");
        }
    }
}

// Signal handler for SIGINT
void sigint_handler(int signum) {

    //check if buffer is empty
    if (strcmp(buffer, "") != 0) {
        memset(buffer, 0, BUF_SIZE);
    }

    //send BYE to server when user presses ctrl+c and TCP is used
    if (strcmp(MODE, "tcp") == 0) {
        send(SOCKET, "BYE\n", strlen("BYE\n"), 0);
        printf("\nBYE\n");
        int len = recvfrom(SOCKET, buffer, BUF_SIZE, 0, NULL, NULL);
        if (len < 0) {
            perror("Failed to receive response");
            close(SOCKET);
            exit(1);
        }
        printf("%s\n", buffer);
    }

    //close socket
    if (SOCKET != 0) {
        close(SOCKET);
    }
    fflush(stdout);
    exit(0);
}

int main(int argc, char *argv[]) {
    // Register signal handler
    signal(SIGINT, sigint_handler);

    //check arguments
    int opt;
    char *host = NULL;
    int port = 0;

    // parsing arguments from command line
    while ((opt = getopt(argc, argv, "h:p:m:")) != -1) {
        switch (opt) {
            case 'h':
                host = optarg;
                break;
            case 'p':
                port = atoi(optarg);
                break;
            case 'm':
                MODE = optarg;
                break;
            default: /* '?' */
                fprintf(stderr, "Usage: %s -h host -p port -m mode\n", argv[0]);
                exit(EXIT_FAILURE);
        }
    }

    validate(host, port);
    create_socket();
    (strcmp(MODE, "tcp") == 0) ? tcp_client(host, port) : udp_client(host, port);

    return 0;
}
